tests/test_app.py

```diff
@@ -70,6 +85,154 @@ def get(self):
         return None
 
 
+class OptionsTest(object):
+    def __init__(self, config_json, init_options=None, want_options=None):
+        self.config_json = config_json
+        self.init_options = init_options
+        self.want_options = want_options
+
+    def init(self):
+        self.config_file_old = os.environ.get(CONFIG_JSON)
+        if self.config_json is not None:
+            if len(self.config_json) == 0 or self.config_json[0] == '{':
+                os.environ[CONFIG_JSON] = self.config_json
+            else:
+                os.environ[CONFIG_JSON] = testutils.resource_filename(self.config_json)
+        elif os.environ.get(CONFIG_JSON):
+            del os.environ[CONFIG_JSON]
+
+    def cleanup(self):
+        if self.config_file_old:
+            os.environ[CONFIG_JSON] = self.config_file_old
+        elif os.environ.get(CONFIG_JSON):
+            del os.environ[CONFIG_JSON]
+
+def named_option_pairs_for_test(named_id_option_pairs):
+    return dict(zip(("ids", "params"), zip(*named_id_option_pairs)))
+
+
+@pytest.fixture(**named_option_pairs_for_test([
```

Comment:
This is a readability nightmare. Can we turn these into individual test cases? At very least, set this dictionary up as a global constant. Then use pytest indirect parameters to pass the test parameters from test case to the fixture: https://docs.pytest.org/en/latest/example/parametrize.html#apply-indirect-on-particular-arguments